nemo/collections/nlp/models/duplex_text_normalization/duplex_decoder.py

```diff
@@ -175,9 +192,36 @@ def _infer(
         # Apply the decoding model
         batch = tokenizer(all_inputs, padding=True, return_tensors='pt')
         input_ids = batch['input_ids'].to(self.device)
-        generated_ids = model.generate(input_ids, max_length=model_max_len)
+        outputs = model.generate(input_ids, output_scores=True, return_dict_in_generate=True, max_length=model_max_len)
+        generated_ids, sequence_toks_scores = outputs['sequences'], outputs['scores']
         generated_texts = tokenizer.batch_decode(generated_ids, skip_special_tokens=True)
 
+        # Use covering grammars (if enabled)
+        if self.use_cg:
+            # Compute sequence probabilities
+            sequence_probs = torch.ones(len(all_inputs)).to(self.device)
+            for ix, cur_toks_scores in enumerate(sequence_toks_scores):
+                cur_generated_ids = generated_ids[:, ix + 1].tolist()
+                cur_toks_probs = torch.nn.functional.softmax(cur_toks_scores, dim=-1)
+                # Compute selected_toks_probs
+                selected_toks_probs = []
+                for jx, _id in enumerate(cur_generated_ids):
+                    if _id != self._tokenizer.pad_token_id:
+                        selected_toks_probs.append(cur_toks_probs[jx, _id])
+                    else:
+                        selected_toks_probs.append(1)
+                selected_toks_probs = torch.tensor(selected_toks_probs).to(self.device)
+                sequence_probs *= selected_toks_probs
+
+            # For TN cases where the neural model is not confident, use CGs
+            neural_confidence_threshold = self.neural_confidence_threshold
+            for ix, (_dir, _input, _prob) in enumerate(zip(input_dirs, input_centers, sequence_probs)):
+                if _dir == constants.INST_FORWARD and _prob < neural_confidence_threshold:
+                    if is_url(_input):
+                        _input = _input.replace(' ', '')  # Remove spaces in URLs
+                    cg_outputs = self.cg_normalizer.normalize(text=_input, verbose=False, n_tagged=1)
+                    generated_texts[ix] = list(cg_outputs)[0]
```

Comment:
Did we discussed to add also the fall back to the input? Could you please add this?

Reply:
I guess the cg grammar will always return something (e.g. the original input if it can handle the input) ? I will double-check that. If that is true, I think the fall back to the input is implicitly handled here.

Reply:
I added code for fallback.